File: mobile/ios/OpenU60/Core/Crypto/ZTECompression.swift
import Foundation
import Compression

/// ZLIB decompression for ZTE config files.
/// Handles plain ZLIB, chunked ZLIB (4-byte BE length prefix), and raw deflate.
enum ZTECompression {

    /// Decompress ZTE config payload. Tries plain zlib, chunked, then raw deflate.
    static func decompress(_ data: Data) throws -> Data {
        // Try plain zlib first
        if let result = decompressZlib(data) {
            return result
        }

        // Try chunked format
        if let result = decompressChunked(data) {
            return result
        }

        // Try raw deflate
        if let result = decompressRaw(data) {
            return result
        }

        // Try skipping potential garbage header bytes
        for skip in [2, 4, 8, 16] {
            if data.count > skip, let result = decompressZlib(data.dropFirst(skip)) {
                return result
            }
        }

        throw CompressionError.decompressionFailed
    }

    /// Compress data using ZLIB.
    static func compress(_ data: Data, chunked: Bool = false, chunkSize: Int = 65536) throws -> Data {
        if !chunked {
            guard let result = compressZlib(data) else {
                throw CompressionError.compressionFailed
            }
            return result
        }

        var result = Data()
        var offset = 0
        while offset < data.count {
            let end = min(offset + chunkSize, data.count)
            let chunk = data[offset..<end]
            guard let compressed = compressZlib(Data(chunk)) else {
                throw CompressionError.compressionFailed
            }
            var length = UInt32(compressed.count).bigEndian
            result.append(Data(bytes: &length, count: 4))
            result.append(compressed)
            offset = end
        }
        return result
    }

    // MARK: - Internal

    private static func decompressZlib(_ data: Data) -> Data? {
        performDecompression(data, algorithm: COMPRESSION_ZLIB)
    }

    private static func decompressRaw(_ data: Data) -> Data? {
        performDecompression(data, algorithm: COMPRESSION_ZLIB)
    }

    private static func decompressChunked(_ data: Data) -> Data? {
        var result = Data()
        var offset = 0
        var chunksFound = 0

        while offset + 4 <= data.count {
            let lengthData = data[offset..<(offset + 4)]
            let chunkLength = lengthData.withUnsafeBytes { ptr -> UInt32 in
                ptr.load(as: UInt32.self).bigEndian
            }
            offset += 4

            if chunkLength == 0 { break }
            guard chunkLength <= data.count - offset else { return nil }

            let chunkData = data[offset..<(offset + Int(chunkLength))]
            offset += Int(chunkLength)

            guard let decompressed = decompressZlib(Data(chunkData)) else { return nil }
            result.append(decompressed)
            chunksFound += 1
        }

        return chunksFound > 0 ? result : nil
    }

    private static func performDecompression(_ data: Data, algorithm: compression_algorithm) -> Data? {
        let destinationBufferSize = data.count * 10
        let destinationBuffer = UnsafeMutablePointer<UInt8>.allocate(capacity: destinationBufferSize)
        defer { destinationBuffer.deallocate() }

        let decompressedSize = data.withUnsafeBytes { sourcePtr -> Int in
            guard let baseAddress = sourcePtr.baseAddress else { return 0 }
            return compression_decode_buffer(
                destinationBuffer, destinationBufferSize,
                baseAddress.assumingMemoryBound(to: UInt8.self), data.count,
                nil, algorithm
            )
        }

        guard decompressedSize > 0 else { return nil }

        // If the buffer was fully used, the data might be larger - try again with bigger buffer
        if decompressedSize == destinationBufferSize {
            let largerSize = destinationBufferSize * 4
            let largerBuffer = UnsafeMutablePointer<UInt8>.allocate(capacity: largerSize)
            defer { largerBuffer.deallocate() }

            let size = data.withUnsafeBytes { sourcePtr -> Int in
                guard let baseAddress = sourcePtr.baseAddress else { return 0 }
                return compression_decode_buffer(
                    largerBuffer, largerSize,
                    baseAddress.assumingMemoryBound(to: UInt8.self), data.count,
                    nil, algorithm
                )
            }
            guard size > 0 else { return nil }
            return Data(bytes: largerBuffer, count: size)
        }

        return Data(bytes: destinationBuffer, count: decompressedSize)
    }

    private static func compressZlib(_ data: Data) -> Data? {
        let destinationBufferSize = data.count + 1024
        let destinationBuffer = UnsafeMutablePointer<UInt8>.allocate(capacity: destinationBufferSize)
        defer { destinationBuffer.deallocate() }

        let compressedSize = data.withUnsafeBytes { sourcePtr -> Int in
            guard let baseAddress = sourcePtr.baseAddress else { return 0 }
            return compression_encode_buffer(
                destinationBuffer, destinationBufferSize,
                baseAddress.assumingMemoryBound(to: UInt8.self), data.count,
                nil, COMPRESSION_ZLIB
            )
        }

        guard compressedSize > 0 else { return nil }
        return Data(bytes: destinationBuffer, count: compressedSize)
    }

    enum CompressionError: LocalizedError {
        case decompressionFailed
        case compressionFailed

        var errorDescription: String? {
            switch self {
            case .decompressionFailed: return "Failed to decompress data"
            case .compressionFailed: return "Failed to compress data"
            }
        }
    }
}

File: mobile/android/OpenU60/app/src/main/java/com/openu60/core/crypto/ZTECompression.kt
package com.openu60.core.crypto

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.util.zip.Deflater
import java.util.zip.Inflater

object ZTECompression {

    fun decompress(data: ByteArray): ByteArray {
        // Try plain zlib first
        tryPlainZlib(data)?.let { return it }

        // Try chunked format: repeated [4-byte BE length][zlib data]
        tryChunkedZlib(data)?.let { return it }

        // Try raw deflate (no zlib header)
        tryRawDeflate(data)?.let { return it }

        // Try skipping header bytes
        for (skip in listOf(2, 4, 8, 16)) {
            if (data.size > skip) {
                tryPlainZlib(data.copyOfRange(skip, data.size))?.let { return it }
            }
        }

        throw ZTEConfigCryptoException("Failed to decompress data: not valid ZLIB or chunked ZLIB format")
    }

    fun compress(data: ByteArray, chunked: Boolean = false, chunkSize: Int = 65536): ByteArray {
        if (!chunked) {
            return plainCompress(data)
        }
        val output = ByteArrayOutputStream()
        var offset = 0
        while (offset < data.size) {
            val end = minOf(offset + chunkSize, data.size)
            val chunk = data.copyOfRange(offset, end)
            offset = end
            val compressed = plainCompress(chunk)
            val lenBuf = ByteBuffer.allocate(4)
            lenBuf.putInt(compressed.size)
            output.write(lenBuf.array())
            output.write(compressed)
        }
        return output.toByteArray()
    }

    private fun plainCompress(data: ByteArray): ByteArray {
        val deflater = Deflater()
        deflater.setInput(data)
        deflater.finish()
        val output = ByteArrayOutputStream()
        val buffer = ByteArray(8192)
        while (!deflater.finished()) {
            val count = deflater.deflate(buffer)
            output.write(buffer, 0, count)
        }
        deflater.end()
        return output.toByteArray()
    }

    private fun tryPlainZlib(data: ByteArray): ByteArray? {
        return try {
            val inflater = Inflater()
            inflater.setInput(data)
            val output = ByteArrayOutputStream()
            val buffer = ByteArray(8192)
            while (!inflater.finished()) {
                val count = inflater.inflate(buffer)
                if (count == 0 && inflater.needsInput()) break
                output.write(buffer, 0, count)
            }
            inflater.end()
            val result = output.toByteArray()
            if (result.isEmpty()) null else result
        } catch (_: Exception) {
            null
        }
    }

    private fun tryRawDeflate(data: ByteArray): ByteArray? {
        return try {
            val inflater = Inflater(true)
            inflater.setInput(data)
            val output = ByteArrayOutputStream()
            val buffer = ByteArray(8192)
            while (!inflater.finished()) {
                val count = inflater.inflate(buffer)
                if (count == 0 && inflater.needsInput()) break
                output.write(buffer, 0, count)
            }
            inflater.end()
            val result = output.toByteArray()
            if (result.isEmpty()) null else result
        } catch (_: Exception) {
            null
        }
    }

    private fun tryChunkedZlib(data: ByteArray): ByteArray? {
        return try {
            val result = ByteArrayOutputStream()
            var offset = 0
            var chunksFound = 0
            while (offset < data.size) {
                if (offset + 4 > data.size) break
                val chunkLen = ByteBuffer.wrap(data, offset, 4).int
                offset += 4
                if (chunkLen == 0) break
                if (chunkLen < 0 || chunkLen > data.size - offset) {
                    throw Exception("Invalid chunk length")
                }
                val chunk = data.copyOfRange(offset, offset + chunkLen)
                offset += chunkLen
                val decompressed = tryPlainZlib(chunk) ?: tryRawDeflate(chunk)
                    ?: throw Exception("Failed to decompress chunk")
                result.write(decompressed)
                chunksFound++
            }
            if (chunksFound == 0) null else result.toByteArray()
        } catch (_: Exception) {
            null
        }
    }
}
